modules/partition-replicator/src/main/java/org/apache/ignite/internal/partition/replicator/PartitionReplicaLifecycleManager.java

```diff
@@ -491,27 +512,43 @@ private CompletableFuture<?> createZonePartitionReplicationNode(
-                                zonePartitionId,
-                                raftClient -> new ZonePartitionReplicaListener(
-                                        new ExecutorInclinedRaftCommandRunner(raftClient, partitionOperationsExecutor)),
-                                new FailFastSnapshotStorageFactory(),
-                                stablePeersAndLearners,
-                                raftGroupListener,
-                                raftGroupEventsListener,
-                                busyLock
-                        ).thenCompose(replica -> executeUnderZoneWriteLock(zonePartitionId.zoneId(), () -> {
-                            replicationGroupIds.add(zonePartitionId);
-
-                            var eventParams = new LocalPartitionReplicaEventParameters(zonePartitionId, revision);
-
-                            return fireEvent(LocalPartitionReplicaEvent.AFTER_REPLICA_STARTED, eventParams);
-                        }))
-                        .thenApply(unused -> false);
-            } catch (NodeStoppingException e) {
-                return failedFuture(e);
-            }
+            var eventParams = new LocalPartitionReplicaEventParameters(zonePartitionId, revision);
+
+            return fireEvent(LocalPartitionReplicaEvent.BEFORE_REPLICA_STARTED, eventParams)
+                    .thenCompose(v -> {
+                        try {
+                            return replicaMgr.startReplica(
+                                    zonePartitionId,
+                                    raftClient -> {
+                                        var runner = new ExecutorInclinedRaftCommandRunner(raftClient, partitionOperationsExecutor);
+
+                                        var replicaListener = new ZonePartitionReplicaListener(runner);
+
+                                        listeners.replicaListenerFuture.complete(replicaListener);
+
+                                        return replicaListener;
+                                    },
+                                    new FailFastSnapshotStorageFactory(),
+                                    stablePeersAndLearners,
+                                    raftGroupListener,
+                                    raftGroupEventsListener,
+                                    busyLock
+                            );
+                        } catch (NodeStoppingException e) {
+                            return failedFuture(e);
+                        }
+                    })
+                    .thenCompose(replica -> executeUnderZoneWriteLock(zonePartitionId.zoneId(), () -> {
+                        replicationGroupIds.add(zonePartitionId);
+
+                        return fireEvent(LocalPartitionReplicaEvent.AFTER_REPLICA_STARTED, eventParams);
+                    }))
+                    .whenComplete((v, e) -> {
+                        if (e != null) {
+                            listenersByZonePartitionId.remove(zonePartitionId);
```

Comment:
Is it possible for the future to be removed because of exception prior to TableManager will try to touch it while handling BEFORE_REPLICA_STARTED?

Reply:
Can you please elaborate, I don't understand the question?

Reply:
I mean following scenario:
1. PRLM fires BEFORE_REPLICA_STARTED, tableManager is going call loadTableListenerToZoneReplica
2. An exception occurred, and thus `listenersByZonePartitionId.remove(zonePartitionId)` was called.
3. loadTableListenerToZoneReplica retrieves `listeners` with `listenersByZonePartitionId.get(zonePartitionId);` and gets Null.

Reply:
Thank you for the clarification. I don't really know how to deal with this case correctly. What is the expected behavior? I think that now the node will simply die immediately, because we will have an exception in the meta storage thread and `loadTableListenerToZoneReplica` will not be called at all. Overall, I added this code just as a safety measure and most likely not needed at all. What do you think, shall I remove it?

Reply:
> What is the expected behavior?

I believe that a fail handler should be called. On it's turn, since we have log only FH we will need to proceed with request processing on such node. Obviously, if there's no replica, all user requests will fail with tx timeout waiting for replica readiness.

From our perspective, it seems that we may treat future unavailability in the map as removed after exception. In that case, we may add the [error] message to the log, and skip further table loading processing.

Reply:
I don't think it's any different from the de-facto current behavior: the node will be in an inoperative state, because we disable Watch processing on any error. And the consecutive NPE from accessing a map will be logged anyway. I can add an assertion there instead, but is it worth it?

Reply:
I thought about explicit null check, but you are right, de-facto it'll be the same. Up to you.

Reply:
I'd leave the current approach, an explicit null check will look weird as if we are trying to recover from a state which is impossible to recover anyway